### PR TITLE
Adds three standalone business intelligence widgets to the Open Mercato dashboard, transforming existing analytics into proactive, actionable business insights.

### PR DESCRIPTION
**Pull request by avishkarroopi**
**Business Health Score**
  - Deterministic 0–100 composite score
  - Based on revenue, orders, AOV, and customer acquisition trends
  - Healthy / Watch / Critical status classification
- **Needs Attention**
  - Rule-based detection of significant metric changes
  - Prioritized critical, warning, and positive signals
  - Handles healthy/no-alert state
- **Executive Brief**
  - Deterministic executive-level summary
  - Synthesizes actual period-over-period business metrics
  - No LLM or external service dependency
### Architecture
- Reuses Open Mercato's existing dashboard widget and analytics infrastructure
- Shared deterministic business intelligence utility
- Three independently registered dashboard widgets
- Edge-case handling for missing, null, zero, and invalid comparison data
### Testing
- 33/33 focused business intelligence tests passed
- 353/353 dashboard regression tests passed
- ESLint passed
- TypeScript typecheck passed
### Notes
Manual UI verification was limited because Docker/PostgreSQL/Redis were unavailable in the local environment. Automated verification completed successfully.

---
<a href="https://app.blacksmith.sh/open-mercato/codesmith/open-mercato/pr/5908"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1791222956&installation_model_id=432243&pr_number=5908&repository=open-mercato%2Fopen-mercato&return_to=https%3A%2F%2Fgithub.com%2Fopen-mercato%2Fopen-mercato%2Fpull%2F5908&signature=ab24ff68a282e527620e68dc3b92bea3e5344ba97abe844edd710b8ce54d0bab"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

File: packages/core/src/modules/dashboards/i18n/en.json [hunk 1]
```diff
@@ -66,6 +66,21 @@
   "dashboards.analytics.widgets.aovKpi.description": "Average order value with period comparison",
   "dashboards.analytics.widgets.aovKpi.error": "Failed to load AOV data",
   "dashboards.analytics.widgets.aovKpi.title": "Average Order Value",
+  "dashboards.analytics.widgets.businessHealthScore.description": "Composite business health score based on revenue, orders, AOV, and customer trends",
```
Comment (adeptofvoltron) [inline packages/core/src/modules/dashboards/i18n/en.json:69]: missing translations to other languaes

File: packages/core/src/modules/dashboards/lib/businessIntelligence.ts [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+}
+
+/**
+ * Identifies high-priority operational items requiring attention based on deterministic business rules.
+ * Sorted by severity (critical first, then warning, then positive).
+ */
+export function identifyAttentionItems(metrics: BusinessMetrics): AttentionItem[] {
+  const items: AttentionItem[] = []
+
+  const rev = metrics.revenue?.change
```
Comment (adeptofvoltron) [inline packages/core/src/modules/dashboards/lib/businessIntelligence.ts:195]: please use meaningful variable names.  "revenue" instead of "rev". same issue in next lines

**Comment by adeptofvoltron** [comment]
@avishkarroopi I got doubts if it should work like that, so I will leave that decision to other maintainers.  
For now I will check the code.
Thank you for your first contribution.
please sign missing CLA

**Comment by adeptofvoltron** [comment]
🤖 `om-auto-review-pr` started by @adeptofvoltron at 2026-09-06T11:55:42Z. Other auto-skills will skip this PR until the lock is released.

**Review by adeptofvoltron — CHANGES_REQUESTED** [review]
# 🔍 Code Review: Adds three standalone business intelligence widgets to the Open Mercato dashboard

## 🎯 Summary

This PR adds three opt-in analytics dashboard widgets — Business Health Score, Needs Attention and Executive Brief — to `packages/core/src/modules/dashboards`, backed by a new shared pure-function library at `lib/businessIntelligence.ts`. The reviewed scope is all 14 changed files: the spec, the shared library and its tests, the three widget folders (`widget.ts` / `config.ts` / `widget.client.tsx`), the widget metadata test, and the English locale file.

There is a lot to like here. The decision to keep the scoring, alerting and narrative generation deterministic and LLM-free is the right call for a core dashboard, and the reasoning is documented inline with explicit weights and thresholds. The library is genuinely pure and separately unit-tested (283 lines of tests for 456 lines of logic), the widgets reuse the existing `useWidgetData` / `WidgetDataBatchProvider` / `DateRangeSelect` infrastructure rather than reinventing it, `lazyDashboardWidget` is used correctly, `hydrateSettings` validates untrusted stored settings defensively, all three widgets ship with `defaultEnabled: false` so the blast radius on existing dashboards is nil, and the icons referenced (`activity`, `alert-triangle`, `file-text`) are all present in the Lucide registry. No database schema, API route, event, CLI or exported-contract surface is touched.

The problems are concentrated in three areas: the PR is currently red on the repository's required gate, the localisation story does not hold up (which is also what @adeptofvoltron flagged in-line), and the widgets present fabricated numbers as measurements when their data fetches fail. I have detailed each below with the file and line to change.

## Verdict

❌ **request changes** — The required CI gate is failing on `i18n:check-sync`, which short-circuits typecheck, tests, DS-drift and the repo-wide audit guards so none of them have actually run on this head; the user-facing copy the widgets render is hardcoded English generated inside the shared library and cannot be localised at all; and a failed or permission-denied metric fetch is silently converted into a neutral value, so the Health Score renders a confident "50 / 100 — Watch" when it in fact has no data. Each of those is independently blocking.

## 🧪 Validation Gate

Validation source: **GitHub checks** — workflow run [33982477413](https://github.com/open-mercato/open-mercato/actions/runs/33982477413) ("CI for Develop&Main") on head `3c95c29`, job `test` (job [101482147811](https://github.com/open-mercato/open-mercato/actions/runs/33982477413/job/101482147811)) — plus a **narrow local fallback** for the gates CI never reached, because the `test` job aborted at step 17 and marked steps 18–26 `skipped`.

| Command | Status | Notes |
|---------|--------|-------|
| `yarn build:packages` | ✅ PASS | CI `prepare` job completed successfully and the `test` job restored its build artifacts. |
| `yarn generate` | ✅ PASS | No auto-discovery drift reported; dashboard widgets are discovered from `widgets/dashboard/*/widget.ts`, and the three new folders follow that convention. |
| `yarn build:packages` (second pass) | ✅ PASS | Same artifacts as above. |
| `yarn i18n:check-sync` | ❌ FAIL | CI step 17 "Check i18n sync" failed. Reproduced locally: `4 issues found across 1 modules` — `[dashboards] de.json / es.json / ko.json / pl.json: 15 missing keys` each. |
| `yarn i18n:check-usage` | ⏭️ NOT RUN | CI step 18 `skipped` — the job aborted at step 17. |
| `yarn typecheck` | ⏭️ NOT RUN | CI step 21 `skipped`. Not reproducible locally in the review worktree without a full `yarn build:packages` first (the failures there are missing `#generated/*` and unbuilt workspace typings across untouched modules, i.e. environment artefacts, not defects in this PR). This gate remains genuinely unverified. |
| `yarn test` | ⏭️ NOT RUN in CI | CI step 22 `skipped`. **Local fallback**: `jest --config packages/core/jest.config.cjs src/modules/dashboards` → **29 suites passed, 353 tests passed**, which confirms the PR description's claim for the dashboards module specifically. The wider workspace suite is still unverified. |
| `yarn build:app` | ⏭️ NOT RUN | CI `docker-build` job `skipped`. |

Also skipped as collateral of the step-17 abort, and therefore unverified on this head: step 24 "Check DS token snapshot drift", step 25 "Check create-app template parity", and step 26 "Run repo-wide audit guards".

Two further check states worth recording, neither of them a code finding:

- `license/cla` is **pending** — the CLA is not signed yet, as @adeptofvoltron already noted. This is a hard merge blocker independent of the code review.
- `mutate` and `prepare` on run [33982338888](https://github.com/open-mercato/open-mercato/actions/runs/33982338888) are still **pending** at the time of writing. Pending checks are not failures and this review does not wait on them; branch protection will hold the merge.
- The blocked preview deployment is expected for a fork PR from a `read`-permission contributor and needs no action from the author.

## Findings

### ⛔ Blocker

**1. `packages/core/src/modules/dashboards/i18n/{de,es,ko,pl}.json` — the four non-English locales are missing all 15 new keys, which fails the required gate and hides everything behind it.**

`en.json` gains 15 keys (lines 69–83 of the diff) and no other locale is updated, so `en.json` now holds 132 keys against 117 in each of `de.json`, `es.json`, `ko.json` and `pl.json`. `yarn i18n:check-sync` is step 17 of the CI `test` job and it fails on exactly this. Because that step aborts the job, steps 18 through 26 — i18n usage, the time-bomb literal check, the agent instruction budget, **typecheck**, **Test**, **Test scripts**, **DS token snapshot drift**, **create-app template parity** and the **repo-wide audit guards** — are all reported as `skipped`. The PR description states that typecheck and 353 tests passed; that may well be true locally, but CI has verified none of it, so this one missing translation set is masking the entire remainder of the gate.

Fix: run `yarn i18n:check-sync --fix` to scaffold the keys, then supply real translations for `de`, `es`, `ko` and `pl` (the `--fix` scaffold alone will satisfy the checker but leaves untranslated values — see `yarn i18n:check-values`). This is also the concrete resolution of @adeptofvoltron's in-line comment on `i18n/en.json:69` ("missing translations to other languaes"). Please push and confirm the `test` job goes green so the skipped gates actually execute.

**2. `packages/core/src/modules/dashboards/lib/businessIntelligence.ts:148–400` — every string the three widgets actually display is hardcoded English generated inside the shared library, so the widgets cannot be localised even once finding 1 is fixed.**

The 15 keys added to the locale files cover only the widget chrome: titles, descriptions, error messages and status labels. All of the substantive copy is produced as literal English sentences by the pure library and rendered verbatim:

- `generateHealthScoreSummary` (lines 154–186) returns sentences such as `Solid business performance driven by ${best.name} growth (${formatPercent(best.change)}).`, plus the metric display names `'Revenue'`, `'Orders'`, `'AOV'` and `'New customers'` (lines 156–159). This is rendered directly at `business-health-score/widget.client.tsx:158`.
- `identifyAttentionItems` (lines 192–318) hardcodes every alert `title` and `description`, for example `'Significant Revenue Contraction'` and `` `Revenue decreased by ${Math.abs(rev).toFixed(1)}% compared with the previous period.` ``. These are rendered at `needs-attention/widget.client.tsx:162` and `:169`.
- `generateExecutiveBrief` (lines 324–400) hardcodes the headlines (`'Business Performance Requires Urgent Attention'`), the narrative fragments and every key-takeaway sentence. Rendered at `executive-brief/widget.client.tsx:118`, `:131` and `:144`.

Root `AGENTS.md` states under **Never**: "Never hard-code user-facing strings — use locale files". A German or Korean user of these widgets would see localised card titles wrapped around English prose. This is a design-level issue rather than a missing-key issue: the pure functions need to return message keys plus interpolation parameters (for example `{ key: 'dashboards.analytics.businessIntelligence.revenueContraction', params: { percent: 12.4 } }`) and let the client components resolve them through `useT()`, rather than returning finished sentences. The library stays pure and deterministic and becomes easier to unit-test on the decision rather than on English phrasing. Please also add the resulting keys to all five locales.

**3. `packages/core/src/modules/dashboards/lib/businessIntelligence.ts:412–448` — a failed or forbidden metric fetch is silently converted into a neutral value, so the widget renders a fabricated score instead of an error.**

`fetchBusinessIntelligenceMetrics` uses `Promise.allSettled` (line 412) and `mapResponse` returns `undefined` for any non-fulfilled result (line 440). `calculateMetricPerformanceMultiplier` then treats a missing trend as the neutral `0.5` multiplier (lines 78–80). Because `Promise.allSettled` never rejects, the `catch` blocks in all three widgets (`business-health-score/widget.client.tsx:68`, `needs-attention/widget.client.tsx:73`, `executive-brief/widget.client.tsx:53`) are unreachable for per-metric failures, so `setError` never fires.

The concrete failure: if all four requests fail — a transient 500, a network blip, or the RBAC denial described in finding 4 — the Health Score widget renders **`50 / 100`** with a **"Watch"** badge and a filled progress bar, which a reader will take as a measurement of their business. Needs Attention renders its reassuring empty state, "All Metrics Within Normal Range". Neither surfaces that nothing was actually fetched. Presenting invented numbers as fact on an executive dashboard is worse than showing nothing.

Fix: have `mapResponse` distinguish "not fetched" from "no comparison data", and have the widgets render an explicit error or unavailable state when some or all metrics failed. A reasonable rule is to surface the error state when any metric is unavailable, or at minimum to refuse to render a score when *all* of them are. Note that the existing "Insufficient comparison data to determine trend trajectory." summary (line 162) is not sufficient — it is a caption underneath a score that has already been displayed as `50`, and the caption itself is wrong: the data was not insufficient, it was never retrieved.

### ⚠️ Major

**4. `business-health-score/widget.ts:11`, `needs-attention/widget.ts:11`, `executive-brief/widget.ts:11` — the widgets declare only `analytics.view` but read two feature-gated entity types, so they are offered to users who cannot retrieve their data.**

All three widgets fetch `sales:orders` and `customers:entities` (`lib/businessIntelligence.ts:413–436`). Those entity types declare `requiredFeatures: ['sales.orders.view']` (`packages/core/src/modules/sales/analytics.ts:7`) and `requiredFeatures: ['customers.view']` (`packages/core/src/modules/customers/analytics.ts:7`), which the batch route enforces server-side via `resolveEntityFeatureAccess` (`lib/widgetDataBatch.ts:26–57`, returning `{ ok: false, error: 'Forbidden' }` at line 79). Every sibling widget in this folder declares its data-source features alongside `analytics.view` — `revenue-kpi`, `orders-kpi`, `aov-kpi`, `orders-by-status`, `revenue-trend` and `sales-by-region` all declare `sales.orders.view`; `new-customers-kpi` declares `customers.people.view`; `top-customers`, which reads both sources exactly as these widgets do, declares all three.

There is no cross-tenant or cross-permission data leak here — the server denies correctly. The consequence is the one described in finding 3: a user holding only `analytics.view` can add all three widgets and will be shown a fabricated `50 / 100` "Watch" score and an "All Metrics Within Normal Range" panel, built entirely from denied requests.

Fix: declare `features: ['analytics.view', 'sales.orders.view', 'customers.view']` on all three widgets. Note this also requires updating the three `features: ['analytics.view']` assertions in `widgets/dashboard/__tests__/businessIntelligenceWidgets.test.ts:30`, `:71` and `:111`.

**5. `needs-attention/widget.client.tsx:34,39,44,141,142,145` and `business-health-score/widget.client.tsx:29,35,41` — hardcoded Tailwind status colours and `dark:` overrides on status colours violate two explicit design-system rules.**

Root `AGENTS.md` states under **Never**: "NEVER use hardcoded Tailwind status colors (`text-red-*`, `bg-green-*`, `text-amber-*`, etc.) — use `{property}-status-{status}-{role}` tokens" and "NEVER add `dark:` overrides on semantic/status tokens — they already handle dark mode". `.ai/ds-rules.md:356` repeats the second rule. The affected code uses `bg-emerald-500` / `bg-amber-500` / `bg-rose-500` for the progress bar, and `border-l-rose-500 bg-rose-50/50 dark:bg-rose-950/20 border-rose-200/60 dark:border-rose-900/40` (and the amber and emerald equivalents) for the severity cards and the empty state.

This is a little unlucky: the CI `ds-lint` job is green, but every `om-ds` rule in `eslint.ds.config.mjs` runs at `warn` during the rollout, so a green `ds-lint` is not evidence of design-system compliance. Separately, the "Check DS token snapshot drift" step never ran on this head because of finding 1.

Fix: use the status tokens the same way the badges in these very files already do (`bg-status-error-bg text-status-error-text` at `needs-attention/widget.client.tsx:35` is correct). For the left accent, `.ai/ds-rules.md:326` gives the exact pattern: `border-l-4 border-status-{status}-border`. Map `critical` → `status-error-*`, `warning` → `status-warning-*`, `positive` → `status-success-*`, and drop every `dark:` variant on those tokens. The `emerald`/`amber`/`rose` names are legitimate as *chart series* colours (`orders-by-status/widget.client.tsx:149` passes them to a chart), but not as entity status colours.

**6. `needs-attention/widget.client.tsx:36,42,46` and `executive-brief/widget.client.tsx:125` — user-visible severity labels bypass i18n, and the Executive Brief renders a raw machine token.**

`SEVERITY_STYLES` hardcodes `badgeText: 'Critical'`, `'Warning'` and `'Growth'`, rendered untranslated at line 166. More visibly, `executive-brief/widget.client.tsx:125` renders `{brief.status}` directly — that is the internal `HealthScoreStatus` union, so the badge displays the lowercase token `healthy`, `watch` or `critical`. The correctly-cased, translatable labels already exist: `dashboards.analytics.widgets.businessHealthScore.statusHealthy` / `statusWatch` / `statusCritical` are in `en.json`, and `business-health-score/widget.client.tsx:109` resolves them properly through `t()`.

Fix: reuse those three existing keys in the Executive Brief badge instead of printing `brief.status`, and route the three severity badge labels through `t()` with new keys added to all five locales.

**7. `lib/businessIntelligence.ts:410` — with `showComparison` disabled, all three widgets silently degrade to a constant, meaningless output while still presenting it as a result.**

`comparison` is `undefined` when `settings.showComparison` is false, so every `MetricTrend.change` is `null` (line 445), every multiplier is the neutral `0.5`, and the score is therefore always exactly `50` with status `watch`. `identifyAttentionItems` returns an empty array because every rule guards on `!= null`, so Needs Attention always shows "All Metrics Within Normal Range", and the Executive Brief always reads "Business Performance Under Watch". The setting is nevertheless offered in all three settings panels (`business-health-score/widget.client.tsx:90–100` and the equivalents), so a user can turn it on and get a permanently fixed, wrong answer.

Fix: either drop the `showComparison` toggle from these three widgets — unlike the KPI cards, they are *defined* by period-over-period comparison and have no meaning without it — or, if it stays, render an explicit "comparison required" state rather than a score.

### 🔹 Minor

**8. `executive-brief/widget.client.tsx:118` vs `:131` — the headline contradicts its own narrative when there is no comparison data.**

With no comparison data the health status is `watch`, so `generateExecutiveBrief` sets the headline to "Business Performance Under Watch" (`lib/businessIntelligence.ts:355–357`) while the narrative immediately below reads "Insufficient comparative metrics available for the selected period to generate an executive brief." (line 362). The widget asserts a business judgement and then states it has no data to judge with. Fix: add a distinct no-data headline for the `takeaways.length === 0` branch instead of falling through to the score-derived one. This is largely a consequence of findings 3 and 7 and may resolve alongside them.

**9. Three files are not valid UTF-8 — they contain Windows-1252 bytes.**

- `lib/businessIntelligence.ts:70` and `:84` contain byte `0xB1` where `±` was intended, and `:102` contains `0x96` where an en dash was intended, rendering as `flat (?2%)` and `Score (0?100)`.
- `lib/__tests__/businessIntelligence.test.ts:20` contains `0xB1` inside a test *name* (`'returns neutral 0.5 for stable trends within ?2%'`), so the mojibake shows up in test output.
- `.ai/specs/2026-09-05-business-intelligence-dashboard-widgets.md:6` and `:29` contain `0x96` in `(0?100)`.

Fix: re-save all three as UTF-8 (or replace with the ASCII equivalents `+/-` and `-`). `businessIntelligenceWidgets.test.ts` is already clean, so this looks like an editor-encoding slip on a subset of files rather than a deliberate choice.

**10. `needs-attention/widget.client.tsx:153,164,178`, `business-health-score/widget.client.tsx:167`, `executive-brief/widget.client.tsx:121,137,157` — arbitrary Tailwind values.**

Root `AGENTS.md` states under **Never**: "NEVER use arbitrary values (`text-[13px]`, `p-[13px]`, `rounded-[24px]`, `z-[9999]`) — use DS scale". The occurrences are `text-[11px]` (four times), `text-[10px]` (twice) and `max-h-[320px]`. Fix: use `text-xs` from the type scale, and a scale-based height such as `max-h-80` for the scroll container.

**11. `needs-attention/widget.client.tsx:164` — `py-0.2` is not a Tailwind spacing value, so no padding is emitted.**

The scale has `py-0` and `py-0.5` but nothing at `0.2`, so this class silently produces no CSS and the severity badge renders with no vertical padding. The sibling badge at `executive-brief/widget.client.tsx:121` correctly uses `py-0.5`. Fix: change it to `py-0.5`.

**12. `lib/businessIntelligence.ts:195–198` — abbreviated variable names (inherited finding).**

Carried forward from @adeptofvoltron's in-line comment of 2026-09-06 on this line ("please use meaningful variable names. `revenue` instead of `rev`. same issue in next lines"), which is still unaddressed on this head. The locals are `rev`, `ord`, `aov` and `cust`, used throughout lines 200–314. Root `AGENTS.md` calls for self-documenting code under **Code Quality**. Fix: rename to `revenueChange`, `orderChange`, `averageOrderValueChange` and `newCustomerChange` — the `...Change` suffix also documents that these are percentages rather than absolute values, which is worth having given the threshold comparisons that follow.

**13. `.ai/specs/2026-09-05-business-intelligence-dashboard-widgets.md` — the spec is missing sections `.ai/specs/AGENTS.md` requires, including the integration-test coverage this repository mandates for new features.**

`.ai/specs/AGENTS.md` requires that "Every non-trivial spec includes: TLDR, Overview, Problem Statement, Proposed Solution, Architecture, Data Models, API Contracts, Risks & Impact Review, Final Compliance Report, Changelog." The spec has Overview, Problem Statement, Proposed Solution, a Security & Access Control section and a Changelog; TLDR, Architecture, Risks & Impact Review and Final Compliance Report are absent. Root `AGENTS.md` additionally requires that "For every new feature, the spec MUST list integration coverage for all affected API paths and key UI paths, and those integration tests MUST ship in the same change" — there is no coverage list and no Playwright spec in this PR.

I am rating this minor rather than major because the change adds no API route and no database surface, and all three widgets are `defaultEnabled: false`, so the UI paths at stake are narrow. It should still be filled in. The Risks section in particular is where findings 3 and 7 would naturally have been caught. Note also that the spec's own claim at line 26 that the shared helper reuses `WidgetDataBatchProvider` is worth restating accurately — the helper consumes the `useWidgetData` fetcher, which is batched only when a provider is present in the tree.

**14. `lib/businessIntelligence.ts:412–437` — three co-installed widgets issue twelve requests for four distinct metrics.**

Each widget calls `fetchBusinessIntelligenceMetrics` independently, and `createWidgetDataBatcher` (`packages/ui/src/backend/dashboard/widgetDataBatcher.ts`) coalesces calls into one HTTP POST but does not deduplicate identical request payloads. A dashboard carrying all three widgets — which is the intended usage — therefore sends 12 batch entries for 4 distinct aggregations, each resolved server-side. The `WidgetDataService` response cache limits the database cost, so this is a real but bounded inefficiency. Fix options, in increasing order of effort: accept it and note it in the spec; hoist the shared fetch into a small context or SWR-style cache keyed on `{dateRange, showComparison}` so the three widgets share one result set; or add payload-level deduplication to the batcher, which would benefit every widget rather than only these three.

### 💅 Nit

**15. `executive-brief/widget.client.tsx:141` — `key={idx}` on the takeaways list.** The array is regenerated wholesale on each fetch so this is harmless today, but the takeaway sentences are stable and unique, and using the sentence (or a stable id emitted alongside it, which finding 2's refactor to message keys would give you for free) reads better and is more robust to future reordering.

**16. `lib/businessIntelligence.ts:117` — the `Math.round(weight * multiplier * 10) / 10` rounding is applied per metric and then again to the total at line 128.** The double rounding is harmless at these magnitudes, but a brief comment explaining that `breakdown[].earnedScore` is deliberately rounded for display while `score` is rounded independently would save the next reader a minute.

## 💥 Breaking Changes

- [x] No exported/public symbol removed or renamed without a deprecation path — the change is purely additive; all 14 files are new except `i18n/en.json`, which only gains keys.
- [x] No function signature changed in a breaking way.
- [x] No required type field removed or narrowed.
- [x] No HTTP route URL removed or renamed; no method changed for an existing operation — no API route is touched.
- [x] No field removed or retyped in an existing response shape.
- [x] No event or message name renamed or removed; no payload field removed.
- [x] No CLI command or flag renamed or removed; no machine-parsed output format changed.
- [x] No database table or column renamed or removed; no column type narrowed — no entity or migration is touched.
- [x] No config key renamed and no default changed silently.
- [x] Where a contract had to change: not applicable, no contract changed.

The three new widget ids (`dashboards.analytics.businessHealthScore`, `dashboards.analytics.needsAttention`, `dashboards.analytics.executiveBrief`) are new additive entries in the dashboard-widget registry and do not collide with any existing id. Per `BACKWARD_COMPATIBILITY.md` this is an additive-only change to a registry surface, which is permitted. Note that if finding 4 is applied, the `features` array on a widget id is a metadata change on a surface no third party has consumed yet, so it is safe to change now — but it would not be after release.

## 🧪 Test Coverage

**What is covered.** `lib/__tests__/businessIntelligence.test.ts` (283 lines) exercises the pure library well: `calculateMetricPerformanceMultiplier` across the neutral band, the positive and negative ramps and the clamps; `calculateHealthScore` across the healthy / watch / critical thresholds including the weighted breakdown; `identifyAttentionItems` across each of the six rules and their severity ordering; and `generateExecutiveBrief` including the no-data branch. `widgets/dashboard/__tests__/businessIntelligenceWidgets.test.ts` (143 lines) covers widget metadata and settings hydration/dehydration for all three widgets, including the invalid-input fallbacks. Locally, `jest --config packages/core/jest.config.cjs src/modules/dashboards` runs 29 suites and 353 tests green, so this coverage does hold — it simply has not run in CI on this head.

**Gaps, with the tests to add.** The most important gap maps directly onto blocker 3: **there is no test for `fetchBusinessIntelligenceMetrics` at all**, which is the one function in the library that is not pure and the one whose failure mode is user-visible. In `lib/__tests__/businessIntelligence.test.ts`, add cases with a stubbed `WidgetDataFetcher` that cover (a) all four requests resolving, asserting the `MetricTrend` mapping including `previous`, `change` and `direction`; (b) one request rejecting, asserting that the caller can distinguish the unavailable metric from a metric with no comparison data; (c) all four rejecting, asserting the same; and (d) `showComparison: false`, asserting whatever contract you settle on for finding 7 rather than the current silent `50`. Second, once findings 4 and 6 are applied, the metadata assertions at `businessIntelligenceWidgets.test.ts:30`, `:71` and `:111` need updating to the corrected `features` arrays — please make them assert the full array rather than a subset so a future regression is caught. Third, there is no rendering test for any of the three `widget.client.tsx` files, so the error, loading, empty and populated states are entirely unverified; a React Testing Library test per widget asserting that a failed fetch renders the error state rather than a score would directly pin blocker 3. Finally, per root `AGENTS.md` and `.ai/qa/AGENTS.md`, a new feature should ship integration coverage for its key UI paths — adding the widget to a dashboard, seeing it render, and toggling its date range — in the same change.

---

*Reviewed by `om-auto-review-pr`. autofix: skipped (not my PR — re-run with `--autofix` to fix it here).*

**Comment by adeptofvoltron** [comment]
🏷️ label rationale

- `changes-requested` (pipeline) — the review verdict is request-changes, driven by three blockers: the failing `i18n:check-sync` gate, hardcoded English copy generated inside `lib/businessIntelligence.ts`, and the silent conversion of failed metric fetches into a fabricated `50 / 100` score. Replaces `review`.
- `feature` (category) — the change adds three new opt-in dashboard widgets plus a shared library; it is not a bug fix or a refactor.
- `priority-medium` (urgency) — a first-time external contribution adding new, `defaultEnabled: false` widgets. Nothing existing is degraded and no user is currently blocked, so this is not high urgency; it is not low either, because the PR currently leaves the repository's required `test` job red on its own head.
- `risk-low` (blast radius) — purely additive: no database schema, migration, API route, event, CLI or exported-contract surface is touched, and all three widgets ship disabled by default, so a merge could not affect an existing dashboard without an explicit user action.
- `needs-qa` (meta) — the change adds UI-rendering `.tsx` files under `packages/core/src/modules/dashboards/widgets/`, so the automated-verification exemption in `AGENTS.md` does not apply. Manual QA is required before merge; per the QA-approval gate this PR must not merge without `qa-approved`. Note the `qa` pipeline label is deliberately not set — that belongs to a QA reviewer once QA is actually in progress.

Not applied: `skip-qa` (would contradict `needs-qa`), and `qa-approved` / `qa-self-verified` (not earned — no manual QA was performed in this run).

**Comment by adeptofvoltron** [comment]
🤖 Handing this back to @avishkarroopi — changes requested.

Thank you again for the contribution; the deterministic, LLM-free approach here is the right instinct for a core dashboard, and the pure library with its own unit tests is nicely structured. The full review is in the `CHANGES_REQUESTED` review above. Three items block merge:

1. **The required CI gate is red.** `yarn i18n:check-sync` fails because `de.json`, `es.json`, `ko.json` and `pl.json` are each missing the 15 new keys. That failure aborts the `test` job at step 17, so typecheck, the test suite, the DS-token drift check, the create-app template parity check and the repo-wide audit guards are all reported `skipped` and have never run on this head. Fixing this is the highest-value single change, because it unblocks verification of everything else. Start with `yarn i18n:check-sync --fix`, then supply real translations. This is also the resolution of the in-line comment already left on `i18n/en.json`.
2. **The widget copy cannot be localised.** All the substantive text — health-score summaries, alert titles and descriptions, executive headlines and takeaways — is generated as hardcoded English sentences inside `lib/businessIntelligence.ts`. The library needs to return message keys plus parameters and let the client components resolve them through `useT()`.
3. **Failed data fetches render as a fabricated score.** `Promise.allSettled` plus a `undefined`-returning mapper means the widgets' `catch` blocks never fire, and missing metrics fall back to a neutral multiplier, so a total fetch failure displays a confident `50 / 100 — Watch`. The widgets need to surface an unavailable state instead.

Beyond those, the review flags four major items (under-declared ACL features versus every sibling widget, hardcoded Tailwind status colours with `dark:` overrides, untranslated severity badges plus a raw `brief.status` token rendered in the Executive Brief, and the `showComparison: false` path silently producing a constant result) and several minors, each with the file, line and a concrete fix. The in-line request to rename `rev` / `ord` / `aov` / `cust` is carried forward as finding 12.

Two things outside the code review: the **CLA is still unsigned**, which is a hard merge blocker regardless of the code, and this PR carries `needs-qa`, so it will also need manual QA sign-off (`qa-approved`) before it can merge. The blocked preview deployment is expected for a fork PR and needs nothing from you.

No changes were pushed to your branch — this run is review-only on another author's PR.

**Comment by adeptofvoltron** [comment]
🤖 `om-auto-review-pr` completed: CHANGES REQUESTED. Lock released.

Full review submitted as a `CHANGES_REQUESTED` review: 3 blockers (failing `i18n:check-sync` gate that skips typecheck/tests/DS-drift/audit guards, non-localisable hardcoded copy in `lib/businessIntelligence.ts`, and fabricated scores on failed metric fetches), 4 majors, 7 minors and 2 nits, including the two in-line comments carried forward. Labels set to `changes-requested` + `feature` + `priority-medium` + `risk-low` + `needs-qa`, and the PR reassigned to @avishkarroopi.

autofix: skipped (not my PR — re-run with `--autofix` to fix it here). No CI follow-up is scheduled: the `test` job has already failed definitively on this head, so the next signal will come from the author's next push.
